Update Lisk framework docs - Closes #769 (#777)
LiskHQ/lisk-docs: modules/ROOT/pages/explanations/architecture/on-chain.adoc

= On-chain architecture
Mona Bärenfänger <mona@lightcurve.io>
//Settings
:toc:
:v_core: 3.0.0

> :imagesdir: ../../assets/images

Reviewer: ```suggestion
:imagesdir: ../../../assets/images
```

:url_github_keys: https://github.com/LiskHQ/lisk-sdk/tree/v5.0.0-alpha.3/framework/src/modules/keys
:url_github_sequence: https://github.com/LiskHQ/lisk-sdk/tree/v5.0.0-alpha.3/framework/src/modules/sequence
:url_github_token: https://github.com/LiskHQ/lisk-sdk/tree/v5.0.0-alpha.3/framework/src/modules/token
// Project URLs
:url_references_framework_basemodule: references/lisk-framework/index.adoc#the-basemodule
:url_references_framework_baseasset: references/lisk-framework/index.adoc#the-baseasset

image::on-chain-architecture.png[]

== What is a custom module?

Custom modules bundle block processing logic and transaction processing logic together with a specific account schema.

They are responsible for the state machine part of the blockchain logic.
The state machine validates and updates the information on either account state or chain state.
Those state changes can be introduced through a block header or a transaction.

Each custom module defines its own data.
If a module is required to interact with other data which is managed by another module, it should communicate through the reducer functions exposed by other modules.

> All of the logic implemented using a custom module / asset must be “deterministic” and executable within block time.

Reviewer: ```suggestion
All of the logic implemented using a custom module / asset must be “deterministic” and executable within the block time.
```

.When to create a custom module
[NOTE]
====

> Module enable to...

Reviewer: ```suggestion
Modules enable the user to perform the following functions:
```

* define how data is stored on the blockchain

> * define logic which is executed per block

Reviewer: ```suggestion
* define the logic which is executed per block
```

> * define logic which is executed per transaction

Reviewer: ```suggestion
* define the logic which is executed per transaction
```

> Also, all of the logic implemented using a module / asset must be “deterministic” and executable within block time.

Reviewer: ```suggestion
In addition, all of the logic implemented using a module / asset must be “deterministic” and executable within the block time.
```

====

=== Custom assets

Custom assets contain all logic related to transactions that belong to the module.
Formerly, this logic was implemented in a custom transaction.
The implementation of a custom asset is similar to a custom transaction, but much more simple.

.When to create a custom asset
[NOTE]
====
Create a custom asset for every transaction type that you want to use in the blockchain application.

> Assets enable to...

Reviewer: ```suggestion
Assets enable users to perform the following:
```

* define a schema for data sent through transaction asset
* validate the data

> * define logic which is executed per asset

Reviewer: ```suggestion
* define the logic which is executed per asset
```

====

== How to create a custom module with an asset

The module and asset must extend the base module and base asset defined in the Lisk SDK.

> === Life cycle of a block

Reviewer: ```suggestion
=== The life cycle of a block:
```

----

<1> `name`(required): used for UI purpose.
<2> `id`(required): used as `AssetID` when sending a transaction.
<3> `schema`(required): defines the `asset` schema for the transaction.
<4> `validate` is used to validate the asset data before it is applied.
Throws an error, in case the validation fails.
<5> `apply`(required): defines a state change induced by this asset.
In `HelloAsset`, it adds the hello string that was sent in the transaction to the senders account and increments the `helloCounter`.

== SDK default modules

[cols="30,70",options="header",stripes="hover"]
|===
|Name
|Description

|DPoS module
a|
The DPoS module is responsible for handling all DPoS related logics.
Specifically:

* Snapshotting vote weights
* Calculating productivity
* Handling registerDelegate, voteDelegate, unlockToken and reportDelegateMisbehavior transaction assets
* Setting the next delegates set

|Keys module
a|
The Keys module handles all logic related to the signatures.

It should verify the signatures based on the multi-signature rules including non-multi-signature accounts.
It also handles the registration of multi-signature accounts.

|Sequence module
a|

> The Sequence module handles all logic related to nonce.

Reviewer: ```suggestion
The Sequence module handles all logic related to a nonce.
```
